marcello3d/trimerge-sync
# Introduce TrimergeStore and TrimergeClient

> I think with a centralized server, sync is pretty simple. Just need an autoincrement version per doc. But with peer to peer (eg local offline sync) it’s a little trickier. Maybe there’s a hybrid model though 
\- me, yesterday

## DATABASE IDEA

What if we separated the notion of a trimerge-sync store and client. You could have multiple clients to one store, and stores can be synced to each other.

### Spitballing here:
* Store:
    * TABLE sync table:
        * sync id: number
        * added: TIMESTAMP
        * refs: string[]
        * snapshotRefs: string[]?
    * TABLE graph nodes:
        * ref: string
        * created: TIMESTAMP
        * baseRef?: string
        * mergeRef?: string
        * editMetadata: JSON
        * delta: JSON
    * TABLE snapshots:
        * ref: string
        * snapshot: JSON
    * API
        * getSyncEvents(since?: number | string, withNodes: boolean)
            * selects from sync table
        * insertSyncEvents(nodes: Node[])
        * insertSnapshot()?
        * getGraphNodes(ref: string[])
            * selects from graph node table
        * getSnapshot(ref: string) 
* Each client:
    * connects to the local database
    * subscribes to ‘sync’ table
        * so all added entries are loaded into memory automatically
    * loads some number of graph nodes into memory 
        * maybe all of them for starters
    * gets the “latest” snapshot
    * determines if any merging needs to happen 
        * how do we know there aren’t older nodes that aren’t loaded that need to be merged?
    * when an edit happens
        * add to local graph synchronously
        * if there are multiple heads, resolve that first
        * writes to the sync table (async)
        * should broadcast to other connected clients—e.g. those subscribed
    * when receiving new syncs
        * add to local graph synchronously
        * if there are multiple heads: 
            * trimerge
            * push sync up to cloud
* When synchronizing a remote database, it’s similar to the client -> database behavior:
    * Need another REMOTES table:
        * url: string (maybe?)
        * remote id: string
        * lastSyncId: number
    * Similar to client logic:
    * Subscribe to the remote database to receive sync events
        * These then get inserted into the local database as new sync events (with new sync ids based on local sync counter)
    * If initializing from blank:
        * Get a snapshot and whatever graph nodes after that are needed to show the latest version of the doc (all merges should have happened already—or very soon after)
    * Whenever new events come in, push them out to the remote

## Open questions:
* How do we prevent thrashing when multiple clients trimerge
* Who generates snapshots and when?
* How do we reduce how much graph needs to be loaded from memory?
* How do we decide how much graph needs to be loaded when initializing from a remote?
    * In theory a database can only be added to from nodes it has already, so I think we can load very minimal amount of the graph when initializing
    * In the case of trimerges, the user that needs to trimerge a big chunk should still have that much of the graph loaded locally
